### PR TITLE
fix(storage): handle delete range in same epoch as write

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Fix https://github.com/risingwavelabs/risingwave/issues/12650

Currently in the code of compaction runner and imm merge, we introduce an assumption that `iter_key.epoch < del_iter.earliest_delete_since(iter_key.epoch)` in a `debug_assert`. However, `del_iter.earliest_delete_since(iter_key.epoch)` returns an epoch that is `>= iter_key.epoch`. Therefore, the `debug_assert` depends on the assumption that the `==` case will never happen, which means that delete range will not delete a key written in the same epoch. However, in log store, we will write the log item to storage when the buffer is full, and after the log item is consumed by the sink, it will truncate the log with a range delete, and both operation happen in the same epoch, which breaks the assumption.

In this PR, we change our code to handle the case when the epoch of iter key is the same as the delete range. Previous when `iter_key.epoch < del_iter.earliest_delete_since(iter_key.epoch)`, it writes an extra delete in the epoch of the range delete. In this PR, in the new case that `iter_key.epoch == del_iter.earliest_delete_since(iter_key.epoch)`, we turn the insert to `HummockValue::Delete` anyway.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/src/hummock/compactor/compactor_runner.rs [hunk 1]
```diff
@@ -797,27 +797,29 @@ where
...
+                    .await?;
+                last_table_stats.total_key_count += 1;
+                last_table_stats.total_key_size += iter_key.encoded_len() as i64;
+                last_table_stats.total_value_size += 1;
+                iter_key.epoch = epoch;
+                is_new_user_key = false;
+            } else {
+                // If the range delete comes from the same epoch, convert value to Delete anyway.
+                debug_assert_eq!(iter_key.epoch, earliest_range_delete_which_can_see_iter_key);
+                value = HummockValue::Delete;
```
Comment (hzxa21) [inline src/storage/src/hummock/compactor/compactor_runner.rs:821]: minor: should we follow the logic in L780-790 to update `last_table_stats` as well?

**Review by hzxa21 — APPROVED** [review]
Generally LGM. @Little-Wallace PTAL

**Comment by Little-Wallace** [comment]
I do not understand.
According to my understanding, If a delete-rand and put key write in the same barrier, this put key shall be filtered.

**Comment by Little-Wallace** [comment]
Why log store can not filter put-key just like `hummock::utils::filter_with_delete_range`?

**Comment by Little-Wallace** [comment]
It is dangerous to keep the same version of put-key with overlapped delete-range in one sstable. Because all our design is based on there will be no key overlap with delete-ranges in the same file.

**Comment by wenym1** [comment]
After offline discussion, we decide to change to truncate the current epoch in the next epoch to avoid doing range delete on a key written in the same epoch.

**Pull request closed without merging**